File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0) return 0;
        int result = 0;
        HashMap<Character, Integer> map = new HashMap<>();

        int left = 0, right = 1;
        map.put(s.charAt(0), 0);
        result = 1;

        while (right < s.length()) {
            if (map.get(s.charAt(right)) == null) {
                map.put(s.charAt(right), right);
                right++;
                result = Math.max(result, right-left);
            } else {
                int index = map.get(s.charAt(right));
                for (int i = left; i <= index; i++) {
                    map.remove(s.charAt(i));
                }
                left = index + 1;
                map.put(s.charAt(right), right);
                result = Math.max(result, right-left);

                right++;
            }
        }
        return result;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        letters = []
        left = 0
        right = 0
        length = 0

        while right < len(s):
            print("test: " + s[right] + ", " + str(right))
            if s[right] in letters:
                if right - left > length:
                    length = right - left
                
                i = left
                while True:
                    if letters[0] != s[right]:
                        del letters[0]
                    else:
                        del letters[0]
                        letters.append(s[right])
                        left = i + 1
                        break
                    i += 1
                right += 1
            else:
                letters.append(s[right])
                right += 1
                if right == len(s):
                    if right - left > length:
                        length = right - left 
        return length
        